net-libs/libwebsockets: update to 4.0.4 and extend configuration options

jsmolic opened the pull request:
Package-Manager: Portage-2.3.89, Repoman-2.3.20
Signed-off-by: Jakov Smolic <jakov.smolic@sartura.hr>
Signed-off-by: Luka Perkov <luka.perkov@sartura.hr>

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -12,9 +12,8 @@ SRC_URI="https://github.com/warmcat/libwebsockets/archive/v${PV}.tar.gz -> ${P}.
-IUSE="access-log caps cgi client dbus generic-sessions http-proxy http2 ipv6
-	+lejp libev libevent libressl libuv peer-limits server-status smtp socks5
-	sqlite3 ssl static-libs threads zip"
+IUSE="access-log caps cgi client dbus +dir +file-ops generic-sessions http-proxy +http-uncommon-headers http2 ipv6 +lejp libev libevent libressl libuv peer-limits server-status smtp socks5 sqlite3 ssl static-libs +sys-capability testapps threads zip"
```
+testapps.

jkroonza replied:
This like is annoying long.  I'm not actually sure what the policy is but consider breaking this up please.

I'm assuming sys-capability is this one:

caps - Use Linux capabilities library to control privilege.

I'm also inclined to say unless it pulls in additional dependencies (which I don't see that you've modified except for mbedtls).

There are others too.  Please always first look for use flags from ${PORTDIR}/profiles/use.desc for use.  Make sure they're the right ones.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -12,9 +12,8 @@ SRC_URI="https://github.com/warmcat/libwebsockets/archive/v${PV}.tar.gz -> ${P}.
-	+lejp libev libevent libressl libuv peer-limits server-status smtp socks5
-	sqlite3 ssl static-libs threads zip"
+IUSE="access-log caps cgi client dbus +dir +file-ops generic-sessions http-proxy +http-uncommon-headers http2 ipv6 +lejp libev libevent libressl libuv peer-limits server-status smtp socks5 sqlite3 ssl static-libs +sys-capability testapps threads zip"
+IUSE+=" ssl_ca_certs ssl_libressl ssl_mbedtls +ssl_openssl"
```
Just merge this with the previous IUSE, *however*:

drop ssl_openssl, this is represented by USE="ssl -libressl -mbedtls"

Rename ssl_ca_certs - the name is counter-intuitive, this really should just be the default with no flag to change it, other than via configuration, ie, if a program doesn't explicitly not use the OS CA certs, use the OS installed CA certs.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -32,8 +31,9 @@ RDEPEND="
 	ssl? (
-		!libressl? ( dev-libs/openssl:0= )
-		libressl?  ( dev-libs/libressl:0= )
+		ssl_libressl?  ( dev-libs/libressl:0= )
+		ssl_mbedtls? ( >net-libs/mbedtls-2.0.0 )
+		ssl_openssl? ( dev-libs/openssl:0= )
```
Adjust as per previous comments:

ssl? (
   libressl? ( ... )
   mbedtls? ( ... )
   !libressl? ( !mbedtls? ( ... ))
)

Just check syntax for the last line please.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -12,9 +12,8 @@ SRC_URI="https://github.com/warmcat/libwebsockets/archive/v${PV}.tar.gz -> ${P}.
+IUSE="access-log caps cgi client dbus +dir +file-ops generic-sessions http-proxy +http-uncommon-headers http2 ipv6 +lejp libev libevent libressl libuv peer-limits server-status smtp socks5 sqlite3 ssl static-libs +sys-capability testapps threads zip"
+IUSE+=" ssl_ca_certs ssl_libressl ssl_mbedtls +ssl_openssl"
 
 REQUIRED_USE="
```
You need to amend REQUIRED_USE to indicate MAX ONE OF mbedtls and libressl.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -47,20 +47,26 @@ src_configure() {
+		-LWS_HAVE_SYS_CAPABILITY_H=$(usex sys-capability)
 		-DLWS_IPV6=$(usex ipv6)
 		-DLWS_ROLE_DBUS=$(usex dbus)
+		-DLWS_SSL_CLIENT_USE_OS_CA_CERTS=$(usex ssl_ca_certs)
```
Default ON this please.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -47,20 +47,26 @@ src_configure() {
 		-DLWS_WITH_GENERIC_SESSIONS=$(usex generic-sessions)
 		-DLWS_WITH_HTTP2=$(usex http2)
 		-DLWS_WITH_HTTP_PROXY=$(usex http-proxy)
+		-DLWS_WITH_HTTP_UNCOMMON_HEADERS=$(usex http-uncommon-headers)
```
Is there a reason to make this optional?  The fact that you're using mbedtls makes me thing you're doing this for embedded work - in which case, yes it makes sense to be able to strip out as much as possible.

jsmolic replied:
Yes, this is for embedded work, hence these options

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -47,20 +47,26 @@ src_configure() {
 		-DLWS_WITH_LIBEV=$(usex libev)
 		-DLWS_WITH_LIBEVENT=$(usex libevent)
 		-DLWS_WITH_LIBUV=$(usex libuv)
+		-DLWS_WITH_MBEDTLS=$(usex ssl_mbedtls)
```
Now that I see this, I don't see explicit LWS_WITH for OPENSSL/LIBRESSL.  Please add those.

jsmolic commented:
@jkroonza Thank you for your review and comments :)\
I fixed it according to your suggestions, I hope everything is good now. If there is something else just comment and I will update accordingly :)

jsmolic commented:
Pushed again, should be okay now:)

jsmolic commented on net-libs/libwebsockets/metadata.xml:
```diff
@@ -19,18 +19,26 @@
+		<flag name="ssl">Include SSL support (defaults to OpenSSL)</flag>
+		<flag name="libressl">Use LibreSSL replacement for OpenSSL</flag>
+		<flag name="mbedtls">Use mbedTLS replacement for OpenSSL</flag>
+		<flag name="testapps">Don't build the libwebsocket-test-apps</flag>
```
Fixed it, I also removed + from `testapps` USE flags since test-apps are disabled by default (since I added `-DLWS_WITHOUT_TESTAPPS=$(usex !testapps)`)

jsmolic commented:
@jkroonza Sorry about this, with `testapps`, `static-libs` USE also has to be set because `LWS_LINK_TESTAPPS_DYNAMIC` is disabled by default. I included new line in REQUIRED_USE which fixes that problem.

jkroonza commented:
> @jkroonza Sorry about this, with `testapps`, `static-libs` USE also has to be set because `LWS_LINK_TESTAPPS_DYNAMIC` is disabled by default. I included new line in REQUIRED_USE which fixes that problem.

No sorry.  Fix it.  static-libs may not be the default, and testapps should be.

jsmolic commented:
@jkroonza It is fixed in https://github.com/gentoo/gentoo/commit/e5b86338747b31607b599102fe9edb685e3265d6

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -12,14 +12,16 @@ SRC_URI="https://github.com/warmcat/libwebsockets/archive/v${PV}.tar.gz -> ${P}.
 	http-proxy? ( client )
 	smtp? ( libuv )
+	ssl? ( ?? ( libressl mbedtls ) )
+	testapps? ( static-libs )
```
No.  testapps must link libraries dynamically.  This is policy.

Sorry, can't let this one slide.  If you need to, engage with Andy.

jkroonza commented on net-libs/libwebsockets/libwebsockets-4.0.1-r1.ebuild:
```diff
@@ -12,14 +12,16 @@ SRC_URI="https://github.com/warmcat/libwebsockets/archive/v${PV}.tar.gz -> ${P}.
-	sqlite3 ssl static-libs threads zip"
+IUSE="access-log caps cgi client dbus +dir +file-ops generic-sessions http-proxy +http-uncommon-headers
+	http2 ipv6 +lejp libev libevent libressl libuv mbedtls peer-limits server-status smtp socks5 sqlite3
+	ssl static-libs testapps threads zip"
```
+testapps

please.  Note +static-libs, see below.

jsmolic commented:
@jkroonza I pushed https://github.com/gentoo/gentoo/commit/d22c3c691b4da1ae4b49ef8bc9eaeb8be334b470 and fixed what you mentioned.\
I set the testapps as default and set dynamic link by default.

However I had some compilation issues with TEST_SERVER_EXTPOLL, and after investigating [this issue](https://github.com/warmcat/libwebsockets/issues/77) and [this one](https://libwebsockets.org/pipermail/libwebsockets/2017-December/003585.html) I added `LWS_WITHOUT_EXTENSIONS` config option, which has to be set along with testapps, so that is also included in this commit.

jsmolic commented:
@jkroonza So, I saw that Andy released 3 new verisons in the past 2 days and I reworked this PR.
I bumped the package to 4.0.4. and dropped 4.0.1.
And I copied new USE flags in the 4.0.4 ebuild in the third commit.

However I'm still having some compilation issues. The one from https://bugs.gentoo.org/719878 doesn't seem to appear.\
With some USE flag combinations (e.g. USE="-*")  I get this:
```
/var/tmp/portage/net-libs/libwebsockets-4.0.4/work/libwebsockets-4.0.4/lib/roles/http/server/server.c: In function 'lws_unauthorised_basic_auth':
/var/tmp/portage/net-libs/libwebsockets-4.0.4/work/libwebsockets-4.0.4/lib/roles/http/server/server.c:873:4: error: 'WSI_TOKEN_HTTP_WWW_AUTHENTICATE' undeclared (first use in this function); did you mean 'WSI_TOKEN_HTTP_CONTENT_RANGE'?
  873 |    WSI_TOKEN_HTTP_WWW_AUTHENTICATE,
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |    WSI_TOKEN_HTTP_CONTENT_RANGE
/var/tmp/portage/net-libs/libwebsockets-4.0.4/work/libwebsockets-4.0.4/lib/roles/http/server/server.c:873:4: note: each undeclared identifier is reported only once for each function it appears in
```


@juippis 
> ```
>  * Messages for package net-libs/libwebsockets-4.0.1-r1:
> 
>  * One or more CMake variables were not used by the project:
>  *   LWS_WITH_LIBRESSL
> ```
I forgot to remove that one, it is fixed in my last commit :)

jsmolic commented:
@jkroonza I checked with Andy about the LWS_WITH_HTTP_UNCOMMON_HEADER problem https://libwebsockets.org/pipermail/libwebsockets/2020-May/008489.html
Per his suggestion I tied LWS_WITH_HTTP_BASIC_AUTH option to the `http-uncommon-header` flag.
That is included in the last https://github.com/gentoo/gentoo/commit/92283efffb5aeb60b994fccad28638145e223da3 commit.

Now the build works fine, I tested with multiple USE flag combinations, and everything seems to be okay
